test: Connection eviction logic tests

mzumsande opened the pull request:
This adds a functional test for the eviction logic for inbound peers, which is triggered when the number of maximum connections is exceeded.

The functional test covers eviction protection for peers that  have sent us blocks or txns recently, or that have faster pings. I couldn't find a way to test the logic of `CConnman::AttemptToEvictConnection` that is based on netgroup (see #14210 for related discussion)

Fixes #16660 (at least partially).

[Edit: Earlier, this PR also contained a unit test, which was removed after the discussion]

mzumsande committed:
Add functional test for P2P eviction logic of inbound peers

instagibbs commented:
TSan build timing out with last output being:
```
Remaining jobs: [feature_block.py, p2p_permissions.py]
```

mzumsande commented:
Hmm, do you think this is caused by my changes? I didn't change the tests that timed out and the changes to `net` were meant to only expose the eviction helper functions for testing (that shouldn't be called in these tests because the code had no previous test coverage).

jamesob commented:
Concept ACK, will review soon

instagibbs commented:
@mzumsande I'll let someone else weigh in on that, I just didn't want to blow away that result in case it was interesting. Restarted that build.

mzumsande commented:
@instagibbs Thanks! It happened again, so it does seem to be caused by the changes here. Will look into it and try to reproduce.

practicalswift commented:
Concept ACK

Thanks for working on this!

mzumsande commented:
Found a few recent travis runs from merges where the TSan build timed out in the same way as here:
https://travis-ci.org/bitcoin/bitcoin/builds/578612449
https://travis-ci.org/bitcoin/bitcoin/builds/578444718
https://travis-ci.org/bitcoin/bitcoin/builds/577975853

naumenkogs commented:
Tested ACK 42a01bf1c15eeb625abdc8bbc0ef9a7852a2b1af

mzumsande commented:
Thanks for the review!
By the way, the functional test will need a small update once #15759 is merged because of the 2 extra outbound connections:
`self.extra_args = [['-maxconnections=32']]` instead of `30`

mzumsande commented:
Rebased and adjusted maxconnections after the merge of #15759 as explained above.

mzumsande commented:
The AppVeyor build failed - there seem to be intermittent failures with the new functional test. I can't reproduce so far, but will look into it.

mzumsande commented:
I think that my assumption that the artificially slowed down peers never have the lowest minimum ping can be wrong, leading to the intermittent failure. I changed the functional test to use `getpeerinfo` RPC to determine what the 8 nodes with the lowest minimum actually are. This should hopefully fix this issue.

jonatack reviewed (commented):
Concept ACK. Nice. Could you mention/describe the proposed changes in net.h/net.cpp in the PR description, and possibly move the combine log/appveyor changes (if they are still needed) to a separate PR.

mzumsande commented:
I added those commits temporarily, because the AppVeyor build was failing and I couldn't reproduce this locally and wanted look at the debug.log, which AppVeyor currently doesnt' show due to #16894. I have removed the commits and fixed my mistake (not giving enough time for ping after adding a new peer).
Will update the PR text as suggested.

jonatack commented:
Thanks, I figured the timing of my review wasn't ideal. Looking forward to reviewing this.

practicalswift commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,135 @@
+""" Test node eviction logic
+
+When the number of peers has reached the limit of maximum connections,
+the next connecting inbound peer will trigger the eviction mechanis.
```
Should be "mechanism" :)

practicalswift commented on src/test/net_tests.cpp:
```diff
@@ -304,4 +305,60 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    // CompareNodeTXTime: erase node with largest nLastTXTime (c5)
+    // in case of a tie, erase node that has fRelayTxes set (c4)
+    // then erase node with fBloomFilter unset (c2)
+    // remaining items are orderes by smallest time connected
```
Should be "ordered" :)

mzumsande replied:
Thanks, fixed both typos.

laanwj commented:
Concept ACK, but I dislike moving internal functionality to `net.h` just for testing.

mzumsande commented:
I don't really see a practical way to avoid that.
I haven't seen any discussions about this in this project, but did some research and the overall consensus on the internet seems to be that internal functionality should not be subject to unit tests (e.g. private members should not be made public just for that).

So would you or others advise that I should just drop the commit with the unit test?

MarcoFalke commented:
> So would you or others advise that I should just drop the commit with the unit test?

@sdaftuar Any conceptual thoughts on this or this pull in general?

sdaftuar commented:
I'm a concept ACK on the p2p test -- that seems great!  However I don't really have an opinion either way on the unit test.

MarcoFalke commented:
@mzumsande Can you split up the functional test, please?

mzumsande commented:
Ok, I removed the unit test from this PR and edited the PR description accordingly.

naumenkogs commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,135 @@
+            block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
+            block.solve()
+            blockpeer.send_blocks_and_test([block], node, success=True)
+            protected.add(len(node.p2ps))
```
Although I understand how it works, this construction takes a little bit of cognitive effort.

Fine for this one, but I suggest using something explicit for the next PRs. Simple `i` would work here, and below, well, you'd have to introduce `i + N(4) + M(5)` or something, but I think it would be easier to comprehend.

mzumsande replied:
I introduced a variable `current_peer`, hopefully that makes it easier to read.

naumenkogs commented:
tested ACK a0780c5ff6818a0a59afd7657bb98c4954ab06b3

MarcoFalke commented:
Open-Close to re-run ci. See https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-487591130

gzhao408 commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,135 @@
+
+class SlowP2PDataStore(P2PDataStore):
+    def on_ping(self, message):
+        time.sleep(0.1)
```
After #18638 this can use `setmocktime` instead of `sleep`?

naumenkogs replied:
I don't think so. Logically this sleep should happen on the mininode side (of couple particular mininodes, not all).
Using mock time would sleep for *all* mininodes, but that doesn't satisfy us.

Unless the mininodes in the test are created in a very particular order, which would probably complicate things.

gzhao408 replied:
Ah, thanks for the info!

jonatack reviewed (commented):
Concept ACK. Tested on the branch and rebased on current master.

Here are some suggestions to improve the test output info/debug logging, move the network setup to a separate function, use `self.setup_clean_chain = False` if you don't need a new chain, import order, clearer naming (`protected_peers` and `evicted_peers`), improve/clarify the assertions, etc. Feel free to use/adapt/ignore: https://github.com/jonatack/bitcoin/commit/c9e96a1

<details><summary>Here is the diff.</summary><p>

```diff
diff --git a/test/functional/p2p_eviction.py b/test/functional/p2p_eviction.py
index b82d644ad7..9bd9872672 100755
--- a/test/functional/p2p_eviction.py
+++ b/test/functional/p2p_eviction.py
@@ -2,16 +2,6 @@
 # Copyright (c) 2019 The Bitcoin Core developers
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-from test_framework.test_framework import BitcoinTestFramework
-from test_framework.mininode import P2PInterface, P2PDataStore
-from test_framework.util import assert_equal, wait_until
-from test_framework.blocktools import create_block, create_coinbase
-from test_framework.messages import CTransaction, FromHex, msg_pong, msg_tx
-
-import time
-
-
 """ Test node eviction logic
 
 When the number of peers has reached the limit of maximum connections,
@@ -26,6 +16,14 @@ Therefore, this test is limited to the remaining protection criteria:
 3.) Protect 8 nodes with the smallest minimum ping
 """
 
+import time
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.mininode import P2PInterface, P2PDataStore
+from test_framework.util import assert_equal, wait_until
+from test_framework.blocktools import create_block, create_coinbase
+from test_framework.messages import CTransaction, FromHex, msg_pong, msg_tx
+
 class SlowP2PDataStore(P2PDataStore):
     def on_ping(self, message):
         time.sleep(0.1)
@@ -36,53 +34,47 @@ class SlowP2PInterface(P2PInterface):
         time.sleep(0.1)
         self.send_message(msg_pong(message.nonce))
 
-
 class P2PEvict(BitcoinTestFramework):
     def set_test_params(self):
-        self.setup_clean_chain = True
+        self.setup_clean_chain = False
         self.num_nodes = 1
-        # The choice of maxconnections results in a maximum of 21 inbound connections
-        # (32 - 10 outbound - 1 feeler). 20 inbounds peers are protected from eviction.
+        # The choice of 32 maxconnections results in a maximum of 21 inbound connections,
+        # e.g. 32 conns - 10 outbound - 1 feeler) = 20 inbound peers protected from eviction.
         self.extra_args = [['-maxconnections=32']]
 
     def setup_network(self):
         self.setup_nodes()
 
-    def run_test(self):
-        protected = set()
+    def setup_test_peers(self):
+        self.protected_peers = set()
         node = self.nodes[0]
         node.generatetoaddress(101, node.get_deterministic_priv_key().address)
 
-        #The first 4 peers send us a block, protecting them from eviction
-        for i in range(4):
-            blockpeer = node.add_p2p_connection(SlowP2PDataStore())
-            blockpeer.sync_with_ping(timeout=5)
+        # The first 4 peers send us a block, protecting them from eviction
+        for _ in range(4):
+            block_peer = node.add_p2p_connection(SlowP2PDataStore())
+            block_peer.sync_with_ping(timeout=5)
             best_block = node.getbestblockhash()
             tip = int(best_block, 16)
             best_block_time = node.getblock(best_block)['time']
             block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
             block.solve()
-            blockpeer.send_blocks_and_test([block], node, success=True)
-            protected.add(len(node.p2ps))
+            block_peer.send_blocks_and_test([block], node, success=True)
+            self.protected_peers.add(len(node.p2ps))
 
-        #The next 5 nodes are slow-pinging peers, making them eviction candidates
-        for i in range(5):
+        # The next 5 nodes are slow-pinging peers, making them eviction candidates
+        for _ in range(5):
             node.add_p2p_connection(SlowP2PInterface())
 
-        #The next 4 peers send us a tx, protecting them from eviction
+        # The next 4 peers send us a tx, protecting them from eviction
         for i in range(4):
             txpeer= node.add_p2p_connection(SlowP2PInterface())
             txpeer.sync_with_ping(timeout=5)
 
             prevtx = node.getblock(node.getblockhash(i+1), 2)['tx'][0]
             rawtx = node.createrawtransaction(
-                inputs=[{
-                    'txid': prevtx['txid'],
-                    'vout': 0
-                }],
-                outputs=[{
-                    node.get_deterministic_priv_key().address: 50 - 0.00125
-                }],
+                inputs=[{'txid': prevtx['txid'], 'vout': 0}],
+                outputs=[{node.get_deterministic_priv_key().address: 50 - 0.00125}],
             )
             sigtx = node.signrawtransactionwithkey(
                 hexstring=rawtx,
@@ -94,18 +86,18 @@ class P2PEvict(BitcoinTestFramework):
                 }],
             )['hex']
             txpeer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
-            protected.add(len(node.p2ps))
+            self.protected_peers.add(len(node.p2ps))
 
         # The next 8 peers have faster pings, which will usually protect them from eviction
-        for i in range(8):
+        for _ in range(8):
             fastpeer = node.add_p2p_connection(P2PInterface())
             wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
 
         # Make sure by asking the node what the actual min pings are
         peerinfo = node.getpeerinfo()
-        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
+        pings = [[0 for __ in range(2)] for _ in range(len(node.p2ps))]
 
-        #After adding 21 peers, the next one hits the maxconnection limit and triggers the eviction mechanism
+        # After adding 21 peers, the next one hits the maxconnection limit and triggers the eviction mechanism
         node.add_p2p_connection(SlowP2PInterface())
 
         # Usually the 8 fast peers are protected. In rare case of unreliable pings,
@@ -115,21 +107,29 @@ class P2PEvict(BitcoinTestFramework):
             pings[i][1] = peerinfo[i]['minping'] if 'minping' in peerinfo[i] else 1000000
 
         sorted_pings = sorted(pings, key = lambda x:x[1])
-
         for i in range(8):
-            protected.add(sorted_pings[i][0])
+            self.protected_peers.add(sorted_pings[i][0])
 
         # One of the non-protected peers must be evicted. We can't be sure which one because
         # 4 peers are protected via netgroup, which is identical for all peers,
-        # and the eviction mechanism doesn't preserve the order of identical elements
-        nodes_disconn = []
+        # and the eviction mechanism doesn't preserve the order of identical elements.
+        self.evicted_peers = []
         for i in range(len(node.p2ps)):
             if(node.p2ps[i].is_connected == False ) :
-                nodes_disconn.append(i)
-        assert_equal(len(nodes_disconn), 1)
-        self.log.info("Protected peers:{}".format(protected))
-        self.log.info("Evicted peer {}".format(nodes_disconn[0]))
-        assert_equal(nodes_disconn[0] in protected, False)
+                self.evicted_peers.append(i)
+
+    def run_test(self):
+        self.log.info("Set up test peer network")
+        self.setup_test_peers()
+
+        self.log.info("Test that only one node was evicted")
+        self.log.debug("{} evicted peers: {}".format(len(self.evicted_peers), set(self.evicted_peers)))
+        assert_equal(len(self.evicted_peers), 1)
+
+        self.log.info("Test that the evicted node is no longer in the set of protected nodes")
+        self.log.debug("{} protected peers: {}".format(len(self.protected_peers), self.protected_peers))
+        assert self.evicted_peers[0] not in self.protected_peers
```
</p></details>

mzumsande commented:
Thanks for the reviews and the suggestions! @jonatack I'll take a close look/push an update early next week!

jonatack commented:
No worries, feel free to ignore :)

I wonder if `protected_peers` is a good name, perhaps just `connected_peers`.

naumenkogs commented:
>I wonder if protected_peers is a good name, perhaps just connected_peers.

I would say all peers by default are connected, so this is not very helpful. Protected is a subset of connected. In addition, protected terminology was used for a long time, so I'd leave it as is.

jonatack commented:
Oh ok, never mind then. Thanks @naumenkogs.

jonatack commented:
> protected terminology was used for a long time

Thanks @naumenkogs, TIL!

mzumsande commented:
@jonatack Thanks again - I took several of your suggestions (some slightly adapted) and pushed an update (just preferred not to split the asserts/logging from the rest)

Yes, `protected_peers` are the subset of peers that are expected not to be evicted by our node (because they sent us a block, tx, have fast pings) - otherwise the test would fail.

naumenkogs commented:
Tested ACK 3758af7

jnewbery commented:
We'll cover this in review club on 2020-06-03: https://bitcoincore.reviews/16756.html

rajarshimaitra commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+        sorted_pings = sorted(pings, key = lambda x:x[1])
+
+        for i in range(8):
+            self.protected_peers.add(sorted_pings[i][0])
```
At this stage, the total number of `protected_peers` is coming out to be 15, which actually should be 16 (4 blocks, 4 txs, 8 fast ping)?
This is probably happening because the previous `protected_peers.add(current_peer)` added the peers with `index + 1`, (`current_peer` is `1` for peer `0`), and `protected_peers.add(sorted_pings[i][0])` is adding them by peer index. This is causing an overlap of the set items and thus one less peer is added to the set. 

Or I might be completely wrong and this is intended behavior.

mzumsande replied:
Nice catch, I introduced this bug in the last push! Will fix later today.

mzumsande replied:
I fixed the initialization of `current_peer`  (so that  having `current_peer=0` for peer `0`).

rajarshimaitra commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        # The choice of 32 maxconnections results in a maximum of 21 inbound connections
+        # (32 - 10 outbound - 1 feeler). 20 inbounds peers are protected from eviction.
```
"20 inbounds peers are protected from eviction." I am not being able to figure the rationale behind this number 20.

andrewtoth replied:
I believe the 20 that are protected are:
4 - netgroup (cannot be tested in this framework)
4 - sent us blocks first
4 - sent us transactions first
8 - lowest ping time
________________
20 - total protected

andrewtoth replied:
Perhaps this would be helpful to clarify in the comments.

mzumsande replied:
Done, I expanded that comment.

rajarshimaitra replied:
Thanks @andrewtoth, that helped a lot.

jonatack replied:
For reference, see `CConnman::AttemptToEvictConnection()` in `src/net.cpp`.

rajarshimaitra commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+            self.protected_peers.add(sorted_pings[i][0])
+
+        # One of the non-protected peers must be evicted. We can't be sure which one because
+        # 4 peers are protected via netgroup, which is identical for all peers,
```
This is coming a little confusing for me, which 4 peers are protected via netgroup? If i understand correctly for functional testing all the peers are from same netgroup. So there's no way to differentiate between them via IP addresses. Is this what this comment is trying to convey?

andrewtoth replied:
I believe these comments above try to convey this behaviour https://github.com/bitcoin/bitcoin/pull/16756/files#diff-96b7763a95e093580bd4507bce07ff22R10-R12

mzumsande replied:
I think it is impossible to say which peers are protected via netgroup because it can vary from run to run! 
Since A) this is the first check in `AttemptToEvictConnection()`  B) all test nodes have identical netgroup and C) `std::sort` does not guarantee to preserve the order of identical elements, this step should protect 4 arbitrary peers in the test. 
If, for example, all 4 block-sending peers would be protected by  chance, then the later `EraseLastKElements` with `CompareNodeBlockTime` would operate on identical elements and erase 4 random peers from `vEvictionCandidates` (protecting them). 
The assertion at the end of the tests only tests that no peer assumed to be protected will be evicted. Its protection might happen in the `EraseLastKElements` specific to its unique feature, or earlier by chance.

jonatack replied:
Yes, the only certainly we have regarding the evicted peers in this test is that it will be one of the zero-indexed peers in the set of {4, 5, 6, 7, 8} e.g. the slow-pinging ones.

rajarshimaitra reviewed (commented):
Concept ACK. Very nice test framework with elaborate comments to understand the logic. Below are few comments for mostly my own understanding and one probable minor bug.

fjahr commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+        for _ in range(5):
+            node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+
```
nit: double new-line

mzumsande replied:
fixed.

fjahr reviewed (approved):
Code review ACK 3758af78564e9a0e059f38c11e941d68be192fb6

Nice test, thank you for adding it!

jonatack commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+        self.log.debug("{} evicted peer: {}".format(len(self.evicted_peers), set(self.evicted_peers)))
+
+        self.log.info("Test that no peer expected to be protected was evicted")
+        assert_equal(self.evicted_peers[0] in self.protected_peers, False)
```
style nit: assertion here is about exclusion, not equality; I think the idiomatic assertion in the tests codebase would be
```suggestion
        assert self.evicted_peers[0] not in self.protected_peers
```

mzumsande replied:
done.

jonatack commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+
+        self.log.info("Test that no peer expected to be protected was evicted")
+        assert_equal(self.evicted_peers[0] in self.protected_peers, False)
+        self.log.debug("{} protected peers: {}".format(len(self.protected_peers), self.protected_peers))
```
Why place the debug logs after the assertions here and line 133? In case of assertion failure, these debug logs would be helpful to see.

mzumsande replied:
done

jonatack commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
```
nit: 2020

jonatack commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+
+    def setup_network(self):
+        self.setup_nodes()
+
```
Unused code?
```diff
-    def setup_network(self):
-        self.setup_nodes()
-
```

jonatack replied:
or better yet, keep it per suggestion in https://github.com/jonatack/bitcoin/commit/c9e96a1 and add:
```diff
    def run_test(self):
+        self.log.info("Set up test peer network")
+        self.setup_test_peers()
```
This provides the important advantage of feedback on how much time is spent in setup while running the test.

mzumsande replied:
The `setup_network` override is not needed imo, removed - thanks!

I added several logging messages for the initial phase (as also suggested below).
 
I am still a bit hesitant with respect to the setup network / test separation, because I don't really view most of it as network setup - in the network setup phase, one node is created, and then in the second phase, we simulate inbound peers connecting to it, sending blocks and transactions, until all slots are taken and a peer gets evicted.
But I will change it if others also think that this is preferable!

jonatack replied:
I'm unsure but suspect one would see `connect_node` statements in the tests if the override was needed.

If you prefer to keep the setup and the assertions all in one block, then no need for `self.protected_peers` and `self.evicted_peers`; that was done in order to separate the network setup from the assertions.

mzumsande replied:
Done, made them local again.

jonatack reviewed (commented):
Almost-ACK. Thanks for taking some of the suggestions in https://github.com/jonatack/bitcoin/commit/c9e96a1.

andrewtoth commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+        node = self.nodes[0]
+        node.generatetoaddress(101, node.get_deterministic_priv_key().address)
+
+        # The first 4 peers send us a block, protecting them from eviction.
```
These single line comments above each block would be more helpful as logs. Rationale https://github.com/bitcoin/bitcoin/blob/master/test/functional/README.md#general-test-writing-advice
```suggestion
        self.log.info("Create the first 4 peers and protect them from eviction by sending us a block.")
```

mzumsande replied:
Done, thanks!

andrewtoth commented:
Tested ACK 3758af78564e9a0e059f38c11e941d68be192fb6

mzumsande commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,141 @@
+            current_peer += 1
+            wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
```
This is probably not material, but the reason is that I wanted to trigger the eviction mechanism immediately after the `getpeerinfo` call, so that the node does not get time to exchange additional pings with its peers while the test does other things.

jnewbery replied:
I wondered the same thing. It's confusing that you do some of the processing before adding the new peer (`pings = [[...`) and some after  (`for i in range(...`). I think you should do all the processing before adding the new connection, since it'll take very little time, and the chance of there being another ping in between is ~zero.

If you do keep this ordering, I think it's worth a comment to say why you're doing it.

mzumsande replied:
I changed the ordering since the probability of furthers pings is extremely low.

rajarshimaitra commented:
Not specific to this PR, but related to the eviction logic code.
https://github.com/bitcoin/bitcoin/blob/657b82cef0e8e8695fc189d013a4353bdbebb041/src/net.cpp#L932-L935
The `naMostConnections` is storing the netgroup key with most number of connections. Why the variable name is like that? First i thought its a typo, but then it seems the `a` is simply added to distinguish it from `nMostConnections`, or is there a specific meaning to it? IMO feels like not so appropriate name to give to a variable which is not storing number of connections.

instagibbs commented:
@rajarshimaitra feel free to open a PR on that if you have ideas on making the code more self-explanatory :+1:

jonatack commented:
ACK 3758af78564e9a0e059f38c11e941d68be192fb6 thanks for adding useful test coverage. Reviewed/tested, re-checked the coherence with `CConnman::AttemptToEvictConnection()`, printed the peers at each step, and added additional assertions and suggestions here: https://github.com/jonatack/bitcoin/commit/e4524a2. I also reckon the code could be shorter if the loops were replaced by lambdas, but simple tests can be better than clever ones.

Thanks also for your patience and taking into consideration all the feedback. I'd enjoy writing more code and documentation, but there are way too many PRs that need review. I wrote additional eviction tests for a follow-up, but it's so hard to get review.

jonatack commented:
@mzumsande thanks for adding the additional logging; it's great.

jnewbery commented:
I have a few style nits which I won't leave inline in order not to distract this PR from functional review:

- L49: at least 2 spaces before inline comments (https://www.python.org/dev/peps/pep-0008/#inline-comments)
- L78: spaces around operators (https://www.python.org/dev/peps/pep-0008/#other-recommendations)
- L114: no spaces around keyword argument (https://www.python.org/dev/peps/pep-0008/#other-recommendations)
- L114: space after colon (https://www.python.org/dev/peps/pep-0008/#pet-peeves)
- L124: prefer `if not expression` over `if expression == false` (https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

Consider installing a python linter like flake8 or black to apply these automatically.

jonatack commented:
Building on John's style points, from `test/functional/README.md`: *When calling RPCs with several arguments, consider using named keyword arguments instead of positional arguments to make the intent of the call clear to readers.*

The variables used as positional args in this test are sufficiently well-named that I didn't bring it up, but it's good to know.

jnewbery commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,136 @@
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
```
Using a list of lists like this is a slightly confusing structure. Consider using a dictionary, and then sorting the `dict.items()` items by value.

mzumsande replied:
Changed it to a dictionary.

jnewbery commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,136 @@
+the same local address. See Issue #14210 for more info.
+
+Therefore, this test is limited to the remaining protection criteria:
+1.) Protect 4 nodes having sent us a block most recently
```
Consider removing these criteria from the file-level doc string and just relying on the comments below. If the criteria ever change, it's likely that this docstring will not get updated and will become outdated.

mzumsande replied:
done

jnewbery reviewed (commented):
Looks great. A bunch of minor suggestions inline.

mzumsande commented:
I just pushed an update that should address all outstanding suggestions from review.

jonatack commented:
Looking at it again, this test can be much simpler. One example: the `current_peer` counter and the `protected_peers` set building can all be removed in favor of just checking that the evicted peer is one of the slow-pinging peers. (After all this review, I may propose the simplifications as a follow-up.)

ACK 45eff751c6d07007dabc365dc4c0e6c63e3fe5cf

naumenkogs commented:
Tested ACK 45eff75

fjahr commented:
re-ACK 45eff751c6d07007dabc365dc4c0e6c63e3fe5cf

Changes since my last reviews were minor improvements addressing nit review comments.

andrewtoth commented on test/functional/p2p_eviction.py:
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
```
nit
```suggestion
# Copyright (c) 2020 The Bitcoin Core developers
```

andrewtoth commented:
re-ACK 45eff751c6d07007dabc365dc4c0e6c63e3fe5cf
